### PR TITLE
Update `SimpleTokenizer` for SAM3 tokenizer convenience

### PR DESCRIPTION
**Pull request by Laughing-q**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds a PyTorch-friendly `__call__` API to the CLIP `SimpleTokenizer` for easy text→token tensor conversion 🚀

### 📊 Key Changes
- Introduced `from __future__ import annotations` for cleaner type hints 🧩
- Added a `torch` dependency in `clip/simple_tokenizer.py` to return tensors 🔥
- Stored commonly used tokenizer constants on init:
  - `sot_token_id` (`<|startoftext|>`)
  - `eot_token_id` (`<|endoftext|>`)
  - default `context_length = 77` 🧠
- Implemented `SimpleTokenizer.__call__(texts, context_length=None) -> torch.LongTensor`:
  - Accepts a single string or list of strings
  - Produces a padded `LongTensor` of shape `[batch, context_length]`
  - Truncates overlong inputs and ensures the last token is `eot_token_id` ✂️

### 🎯 Purpose & Impact
- Makes tokenization easier to use in PyTorch pipelines (call tokenizer directly to get model-ready tensors) ✅
- Standardizes CLIP-like behavior with a default context length of 77, reducing boilerplate 📏
- Improves performance and ergonomics for batching (automatic padding + truncation) ⚡
- Potential impact: introduces a hard dependency on PyTorch for this module; environments without `torch` may need to install it or avoid importing this tokenizer 📦

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Laughing-q, thank you for submitting a `ultralytics/CLIP` 🚀 PR! This is an automated message, and an engineer will assist soon. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/CLIP/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/CLIP` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: clip/simple_tokenizer.py [hunk 1]
```diff
@@ -145,3 +148,27 @@ def decode(self, tokens):
...
+
+        Returns:
+            (torch.Tensor): A two-dimensional tensor containing the resulting tokens, shape = [number of input strings,
+                context_length]
+        """
+        if isinstance(texts, str):
+            texts = [texts]
+        context_length = context_length or self.context_length
+        assert context_length, "Please set a valid context length"
+        all_tokens = [[self.sot_token_id, *self.encode(text), self.eot_token_id] for text in texts]
```
Comment (UltralyticsAssistant) [inline clip/simple_tokenizer.py]: 💡 **MEDIUM**: Using `assert` for runtime input validation is fragile because assertions can be disabled with Python optimizations (`-O`), potentially allowing `context_length=0` to pass silently. Prefer raising a `ValueError` to enforce this invariant reliably.

**Suggested change:**
```suggestion
        if not context_length:
            raise ValueError("Please set a valid context length")
```

File: clip/simple_tokenizer.py [hunk 2]
```diff
@@ -145,3 +148,27 @@ def decode(self, tokens):
...
+        result = torch.zeros(len(all_tokens), context_length, dtype=torch.long)
+        for i, tokens in enumerate(all_tokens):
+            if len(tokens) > context_length:
+                tokens = tokens[:context_length]  # Truncate
+                tokens[-1] = self.eot_token_id
+            result[i, : len(tokens)] = torch.tensor(tokens)
+        return result
```
Comment (UltralyticsAssistant) [inline clip/simple_tokenizer.py]: 📝 **LOW**: `torch.tensor(tokens)` is created without dtype/device, relying on defaults and then assigned into a `torch.long` CPU tensor. This can introduce unnecessary casting/copies and makes device placement harder for callers. Consider constructing the token tensor with `dtype=result.dtype` and `device=result.device` for consistency.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Overall the change is small and the new callable API looks correct for padding/truncation with SOT/EOT wrapping. Main issues: avoid `assert` for runtime validation, fix the malformed docstring, and consider aligning per-row tensor creation dtype/device with the preallocated output tensor to prevent extra casts/copies and improve usability.

💬 Posted 3 inline comments

File: clip/simple_tokenizer.py [hunk 3]
```diff
@@ -145,3 +150,27 @@ def decode(self, tokens):
...
+            texts (Union[str, list[str]]): An input string or a list of input strings to tokenize.
+            context_length (int): The context length to use.
+
+        Returns:
+            (torch.Tensor): A two-dimensional tensor containing the resulting tokens, shape = [number of input strings,
+                context_length]
+        """
+        if isinstance(texts, str):
+            texts = [texts]
+        context_length = context_length or self.context_length
```
Comment (UltralyticsAssistant) [inline clip/simple_tokenizer.py]: ⚠️ **HIGH**: `context_length = context_length or self.context_length` will raise `AttributeError` if `self.context_length` is not defined on this tokenizer instance (it’s not set in the shown `__init__`). If the intent is to have a default, it should be explicitly initialized in `__init__` or safely derived (e.g., via `getattr`) with a clear error when absent.

File: clip/simple_tokenizer.py [hunk 4]
```diff
@@ -145,3 +150,27 @@ def decode(self, tokens):
...
+        assert context_length, "Please set a valid context length"
```
Comment (UltralyticsAssistant) [inline clip/simple_tokenizer.py]: 💡 **MEDIUM**: Using `assert` for input validation can be bypassed with Python optimizations (`-O`), turning this into a potential silent failure later. Prefer an explicit exception (e.g., `ValueError`) for user-facing API validation.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 2

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Clean, focused change overall: caching SOT/EOT IDs and adding a callable API is useful. The main risk is relying on `self.context_length` without guaranteeing it exists, which can cause runtime `AttributeError`. Also, `assert` isn’t ideal for validating public inputs; an explicit exception is safer.

💬 Posted 3 inline comments

File: clip/simple_tokenizer.py [hunk 5]
```diff
@@ -145,3 +151,27 @@ def decode(self, tokens):
         """Decodes a list of BPE tokens into a UTF-8 string, replacing '</w>' with a space."""
         text = "".join([self.decoder[token] for token in tokens])
         return bytearray([self.byte_decoder[c] for c in text]).decode("utf-8", errors="replace").replace("</w>", " ")
+
+    def __call__(self, texts: str | list[str], context_length: int | None = None) -> torch.LongTensor:
```
Comment (UltralyticsAssistant) [inline clip/simple_tokenizer.py:155]: 💡 **MEDIUM**: `context_length = context_length or self.context_length` treats `0` the same as `None`, and the subsequent `assert context_length` relies on truthiness rather than type/range. This can lead to surprising behavior (e.g., `context_length=0` silently falling back) and asserts can be stripped with `-O`. Consider explicit `None` handling and a clear `ValueError` for non-positive lengths.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 3

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Overall clean, focused change: caching SOT/EOT IDs and adding a callable API is straightforward and should help downstream usage. The only issue worth addressing is the `context_length` validation: using `or` plus an `assert` can mask invalid inputs and may be skipped in optimized runs; switching to explicit `None` handling and raising `ValueError` would make this more robust.

💬 Posted 1 inline comment

**Review by fcakyon — APPROVED** [review]

**Comment by UltralyticsAssistant** [comment]
Merged — thank you for the awesome improvement, @Laughing-q (and thanks @fcakyon for the contributions)! 🎉

As **Leonardo da Vinci** famously said, *“Simplicity is the ultimate sophistication.”* This PR embodies that: adding a clean, PyTorch-friendly `SimpleTokenizer.__call__` makes text → model-ready token tensors feel effortless, while the default `context_length=77`, padding, and safe truncation bring CLIP-like ergonomics with far less boilerplate.

Really appreciate the thoughtful design choices and the focus on developer experience—this will make batching and integration into PyTorch pipelines much smoother for everyone.
